overlord/ifacestate/handlers.go

```diff
@@ -485,6 +487,145 @@ func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
+		// go with those from the new core snap.
+		if len(candidates) == 2 {
+			switch {
+			case candidates[0].Snap.Name() == "ubuntu-core" && candidates[1].Snap.Name() == "core":
+				candidates = candidates[1:2]
+			case candidates[1].Snap.Name() == "ubuntu-core" && candidates[0].Snap.Name() == "core":
+				candidates = candidates[0:1]
+			}
+		}
+		if len(candidates) != 1 {
+			crefs := make([]string, len(candidates))
+			for i, candidate := range candidates {
+				crefs[i] = candidate.String()
+			}
+			task.Logf("cannot auto-connect plug %s, candidates found: %s", plug, strings.Join(crefs, ", "))
+			continue
+		}
+		slot := candidates[0]
+		connRef := interfaces.NewConnRef(plug, slot)
+		key := connRef.ID()
+		if _, ok := conns[key]; ok {
+			// Suggested connection already exist so don't clobber it.
+			// NOTE: we don't log anything here as this is a normal and common condition.
+			continue
+		}
+
+		ts, err := AutoConnect(st, chg, plug.Snap.Name(), plug.Name, slot.Snap.Name(), slot.Name)
+		if err != nil {
+			task.Logf("cannot auto-connect plug %s to %s: %s", connRef.PlugRef, connRef.SlotRef, err)
+			continue
+		}
+		autots.AddAll(ts)
+	}
+	// Auto-connect all the slots
+	for _, slot := range m.repo.Slots(snapName) {
+		candidates := m.repo.AutoConnectCandidatePlugs(snapName, slot.Name, autochecker.check)
+		if len(candidates) == 0 {
+			continue
+		}
+
+		for _, plug := range candidates {
+			// make sure slot is the only viable
+			// connection for plug, same check as if we were
+			// considering auto-connections from plug
+			candSlots := m.repo.AutoConnectCandidateSlots(plug.Snap.Name(), plug.Name, autochecker.check)
+
+			if len(candSlots) != 1 || candSlots[0].String() != slot.String() {
+				crefs := make([]string, len(candSlots))
+				for i, candidate := range candSlots {
+					crefs[i] = candidate.String()
+				}
+				task.Logf("cannot auto-connect slot %s to %s, candidates found: %s", slot, plug, strings.Join(crefs, ", "))
+				continue
+			}
+
+			connRef := interfaces.NewConnRef(plug, slot)
+			key := connRef.ID()
+			if _, ok := conns[key]; ok {
+				// Suggested connection already exist so don't clobber it.
+				// NOTE: we don't log anything here as this is a normal and common condition.
+				continue
+			}
+			ts, err := AutoConnect(st, chg, plug.Snap.Name(), plug.Name, slot.Snap.Name(), slot.Name)
+			if err != nil {
+				task.Logf("cannot auto-connect slot %s to %s: %s", connRef.SlotRef, connRef.PlugRef, err)
+				continue
+			}
+			autots.AddAll(ts)
+		}
+	}
+
+	lanes := task.Lanes()
+	if len(lanes) == 1 && lanes[0] == 0 {
+		lanes = nil
+	}
+	ht := task.HaltTasks()
+
+	// add all connect tasks to the change of main "auto-connect" task and to the same lane.
+	for _, l := range lanes {
+		autots.JoinLane(l)
+	}
+	chg.AddAll(autots)
+	// make all halt tasks of the main 'auto-connect' task wait on connect tasks
+	for _, t := range ht {
+		t.WaitAll(autots)
+	}
+
+	st.EnsureBefore(0)
```

Comment:
do we need this? I remember we needed it some place but usually is not needed

Reply:
This is a curious one. Not doing this here makes a lot of ifacestate tests fail with the status of entire change == "Do" instead of "Done", because all the newly added tasks are in "Do" state; no amount of Ensure/Wait (on both ifacemgr and hookmgr instances in test) helps, the best result I've got was all tasks transitioned to "Done", but the state of entire change was stuck on "Do". Seems to be an internal quirk of the task processing that I don't get. I'm happy to address dig into this, any hints?

Reply:
I'm likely missing something here, but if you do Ensure manually  this is not relevant, otoh in normal usage (with a full overlord/taskrunners)  we should hit afaiu:
 
https://github.com/snapcore/snapd/blob/master/overlord/state/taskrunner.go#L238

it's not a priority to understand this though,  more important to get the rest fixed